engine/src/main/java/org/camunda/bpm/engine/impl/bpmn/deployer/BpmnDeployer.java

```diff
@@ -309,32 +309,27 @@ protected void addMessageStartEventSubscription(EventSubscriptionDeclaration mes
 
   }
 
-  protected boolean isSameMessageEventSubscriptionAlreadyPresent(EventSubscriptionDeclaration eventSubscription, String tenantId) {
+  protected boolean isSameMessageStartEventSubscriptionAlreadyPresent(EventSubscriptionDeclaration eventSubscription,
+                                                                      String tenantId) {
     // look for subscriptions for the same name in db:
-    List<EventSubscriptionEntity> subscriptionsForSameMessageName = getEventSubscriptionManager()
-      .findEventSubscriptionsByNameAndTenantId(EventType.MESSAGE.name(), eventSubscription.getUnresolvedEventName(), tenantId);
+    EventSubscriptionEntity subscriptionForSameMessageName = getEventSubscriptionManager()
+        .findMessageStartEventSubscriptionByNameAndTenantId(eventSubscription.getUnresolvedEventName(), tenantId);
 
     // also look for subscriptions created in the session:
     List<EventSubscriptionEntity> cachedSubscriptions = getDbEntityManager()
-      .getCachedEntitiesByType(EventSubscriptionEntity.class);
+        .getCachedEntitiesByType(EventSubscriptionEntity.class);
 
     for (EventSubscriptionEntity cachedSubscription : cachedSubscriptions) {
+      if (eventSubscription.getUnresolvedEventName().equals(cachedSubscription.getEventName()) &&
+          hasTenantId(cachedSubscription, tenantId) &&
+          !subscriptionForSameMessageName.equals(cachedSubscription)) {
 
-      if(eventSubscription.getUnresolvedEventName().equals(cachedSubscription.getEventName())
-        && hasTenantId(cachedSubscription, tenantId)
-        && !subscriptionsForSameMessageName.contains(cachedSubscription)) {
-
-        subscriptionsForSameMessageName.add(cachedSubscription);
+        subscriptionForSameMessageName = cachedSubscription;
+        break;
       }
     }
 
-    // remove subscriptions deleted in the same command
-    subscriptionsForSameMessageName = getDbEntityManager().pruneDeletedEntities(subscriptionsForSameMessageName);
-
-    // remove subscriptions for different type of event (i.e. remove intermediate message event subscriptions)
-    subscriptionsForSameMessageName = filterSubscriptionsOfDifferentType(eventSubscription, subscriptionsForSameMessageName);
```

Comment:
💭 A `cachedSubscription` can be for intermediate message event and we don't filter those after removing this line.
At lines 323-325 we compare only for the `eventNames`, we can consider using `#isSubscriptionOfDifferentTypeAsDeclaration` there.
As there are no CI failure, I assume there's no test coverage for this use case.
If we do remove `#filterSubscriptionsOfDifferentType`, it can be cleaned up as it's used nowhere else.

Reply:
Nice catch! 👍